enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,151 @@
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes to add the support of RAID and BIOS configuration for baremetal IPI deployments.
+
+## Motivation
+
+With the new features of [RAID configuration](https://github.com/metal3-io/baremetal-operator/pull/292) and [BIOS configuration](https://github.com/metal3-io/baremetal-operator/pull/302), openshift already has the ability to configure RAID and BIOS for baremetal.
+
+However, these features only work with a running cluster, users cannot configure desired RAID and BIOS during IPI deployments.
+
+### Goals
+
+- Allow users to manage the RAID and BIOS configuration for control plane hosts during installation.
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. The user runs `create manifests` and then adds RAID and BIOS fields to the **BMH** created by the installer.
+2. The interface to terraform in the installer code parses the BIOS and RAID fields.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for control plane hosts and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Add two fields
+
+The user runs `create manifests` and then adds RAID and BIOS fields to the **BMH** created by the installer.
+
+```yaml
+spec:
+  bmc:
+    address: <bmc>://<ip-address>
+    credentialsName: worker-0-bmc-secret
+  bootMACAddress: mac-address
+  raid: raid-config
+  firmware: bios-config
+  ...
+```
+
+RAID feature has been implemented in **BMO**, so the *raid* field here is the same as the [BMH](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/config/crd/bases/metal3.io_baremetalhosts.yaml).
+The *firmware* field is the same as the *spec.firmware* field in **BMH** which is been advancing by [#302](https://github.com/metal3-io/baremetal-operator/pull/302).
+
+#### Process the fields in installer
+
+For control plane hosts, transform the *raid* and *firmware* fields into json respectively and then write them into the
+***terraform.baremetal.auto.tfvars.json*** file as the values of *raid_config* and *bios_settings* respectively.
+
+#### Process the fields in terraform-provider-ironic
```

Comment:
I'm curious how it will be done. Terraform is known to be bad with nested JSON fields, you may have to pass everything as a string..

(just food for thought, not really a comment on this spec)

Reply:
Sorry for the misunderstanding, yes, we did pass the RAID and BIOS to terraform-provider-ironic in [string form](https://github.com/hs0210/installer/commit/9072a656176309daca0eee595f9587df1eb3c860#diff-3b0c0219b1206253af5d430d464ea73c976c339a055510a0d9456d4dfb0419afR107-R108).